[14.0][MIG] sale promotion rule

xavier-bouquiaux opened the pull request.

bguillot committed:
sale_promotion_rule: module to manage promotion rules

bguillot committed:
sale_promotion_rule: add button to clean promotion rule

lmignon committed:
sale_promotion_rule: Improve implementation
The module now supports multiple promotion rules.
When a coupon is used, it always take precedence on automatic promotion
rule.
The computation of promotions has been refactored to update lines from
the sale.order object. Sine lines are updated from the SO, the recompute
of SO fields that depends of lines is only tiggered at the end of the
update on the lines. (In place of each time a line is updated).
You can ask to recompute the promotion of a sale.order. If a coupon is
already specified it's conserved but all the automatic rules are applied
again.

lmignon committed:
[IMP] sale_promotion_rule: Remove demo data Avoid demo data with introducing a side effect on the expected results of all sale order created into unitest

lmignon committed:
[IMP] sale_promotion_rule: Improve README

lmignon committed:
[IMP] sale_promotion_rule: Supports discount amount
Before this change, only percentage were applicable as discount by
promotion rule.
This change allows you to define rules that apply a discount in the form of a specific amount. This amount is translated as a percentage disctributed on the different lines during the rule resolution in order to benefit from the standard mechanism for calculating discount by Odoo.

lmignon committed:
[IMP] sale_promotion_rule: s/restriction_amount/restriction_amount_field
The field reference the name of the field to use on the sale order to determine if the promotion is applicable according to the total amount.

lmignon committed:
[IMP] sale_promotion_rule: Overrides name_get in place of defining a new field for display_name

lmignon committed:
[IMP] sale_promotion_rule: Improve performance
When checking if a promotion rule can be applied according to its usage definition, usage search_count in place of search since we are not interested by the resultset

lmignon committed:
[FIX] sale_promotion_rule: We must also include Coupon into the multi rule strategy check

[UPD] Update sale_promotion_rule.pot

lmignon committed:
[IMP] sale_promotion_rule: Use discount product for discount defined with fixed amount
While the first approach based on the reuse of the standard Odoo field "discount_amount" seemed the least intrusive, it is unfortunately impossible to avoid rounding problems due to the transformation of a fixed amount into a percentage of dicount to be applied on the different products. Rounding problems are further aggravated when a so contains lines with products with a high price difference.
Since it's not possible to rely on the standard Odoo field, the implementation of the fixed amount discount is now based the usage of a dedicated product.

lmignon committed:
[IMP] sale_promotion_rule: Add hook to exclude lines from the sale order amount total
In some cases some lines must not be taken into account into the sale order total when checking if the minimal amount specified on the promotion rule is reached.
As side effect we must be more strict on the way we get the total on te sale order and if taxes are included or excluded into the total

lmignon committed:
[IMP] sale_promotion_rule: Improve promotion rule form

lmignon committed:
[IMP] sale_promotion_rule: Improve tests coverage

lmignon committed:
[FIX] sale_promotion_rule: Fix search for rules already used for one partner

Commit:
[FIX] prevent to aply an automatic rule on a sale promotion line

OCA-git-bot committed:
[ADD] icon.png

[UPD] README.rst

acsonefho committed:
[10.0][ADD] Add company_id + related rules

acsonefho committed:
[10.0][FIX] Fix sale_promotion_rule in case of multi rules

[UPD] Update sale_promotion_rule.pot

Commit:
[FIX] minimum amount must be evaluate without any discount applied

[UPD] Update sale_promotion_rule.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

[UPD] Update sale_promotion_rule.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

rousseldenis committed:
!fixup

[UPD] Update sale_promotion_rule.pot

[UPD] README.rst

OCA-git-bot committed:
sale_promotion_rule 10.0.2.0.1

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

[UPD] Update sale_promotion_rule.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

OCA-git-bot committed:
sale_promotion_rule 10.0.2.0.2

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

[UPD] Update sale_promotion_rule.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

georginaf committed:
Added translation using Weblate (Catalan)

rousseldenis committed:
[10.0][FIX] sale_promotion_rule: Apply coupon to lines if percentage

If the coupon rule apply a discount in fixed price, the rule should
not be applied on each sale order line

Commit:
[MIG] 12.0 sale_promotion_rule

rousseldenis committed:
[10.0][FIX] sale_promotion_rule: Enable sequence order

Sequence order was never applied du of missing _order

[UPD] Update sale_promotion_rule.pot

OCA-git-bot committed:
sale_promotion_rule 10.0.2.1.0

OCA-git-bot committed:
sale_promotion_rule 10.0.2.1.1

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

[UPD] Update sale_promotion_rule.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_promotion_rule
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_promotion_rule/

qgroulard committed:
[FIX] Clear promotion keep manual discount

If no rule is percentage (i.e. edit discount field) we should not clear the discount when removing promotions.

qgroulard committed:
[IMP] sale_promotion_rule: black, isort, prettier

qgroulard committed:
[MIG] sale_promotion_rule: Migration to 13.0

qgroulard committed:
[IMP] sale_promotion_rule: black, isort, prettier

xavier-bouquiaux committed:
[IMP] sale_promotion_rule : black, isort, prettier

qgroulard committed:
[IMP] Add sale demo data

Fix tests using demo data modified by other modules

xavier-bouquiaux committed:
[WIP][14.0][MIG] sale_promotion_rule

xavier-bouquiaux committed:
[WIP][14.0][MIG] sale_promotion_rule

rousseldenis committed:
[14.0][IMP] sale_promotion_rule: Add test for multi currency case

rousseldenis committed:
[14.0][IMP] sale_promotion_rule: Improve coverage testing wizard

rousseldenis committed:
[14.0][IMP] sale_promotion_rule: Improve coverage testing has_promotion_rules

rousseldenis commented:
@hparfr IMHO, either a Roadmap, either a modification before merge should integrate coupon module that has been split by Odoo from 13 => 14 as it could be interesting to base our flow on coupon.coupon model

https://github.com/odoo/odoo/blob/14.0/addons/coupon/models/coupon.py#L10

Not too much logic there, no mandatory fields. What do you think ?

@sebastienbeau @Cedric-Pigeon @qgroulard

rousseldenis committed:
[14.0][i18n] sale_promotion_rule: Regenerate pot and merge to avoid duplicates

hailangvn commented on sale_promotion_rule/views/sale_promotion_rule.xml:
```diff
@@ -0,0 +1,96 @@
+                            />
+                            <field name="multi_rule_strategy" />
+                        </group>
+                        <group name="condition" string="Condition" col="3">
+                            <label for="date_from" string="Date début/fin" />
+                            <field name="date_from" nolabel="1" />
+                            <field name="date_to" nolabel="1" />
+                            <field name="minimal_amount" colspan="3" />
+                            <field name="is_minimal_amount_tax_incl" colspan="3" />
+                            <field name="only_newsletter" colspan="3" />
+                            <field
+                                name="restrict_pricelist_ids"
+                                widget="many2many_tags"
+                                colspan="3"
+                            />
+                            <label for="restrict_partner_ids" colspan="3" />
+                            <field
+                                name="restrict_partner_ids"
+                                nolabel="1"
+                                colspan="3"
+                            />
+                        </group>
+                    </group>
```
```xml
                    </group>
                    <group name="condition" string="Condition" col="6">
                        <label for="date_from" string="Date début/fin"/>
                        <field name="date_from" nolabel="1"/>
                        <field name="date_to" nolabel="1"/>
                        <field name="minimal_amount" colspan="3"/>
                        <field name="is_minimal_amount_tax_incl" colspan="3"/>
                        <field name="only_newsletter" colspan="3"/>
                        <field name="restrict_pricelist_ids" widget="many2many_tags" colspan="3"/>
                        <label for="restrict_partner_ids" colspan="6"/>
                        <field name="restrict_partner_ids" nolabel="1" colspan="6"/>
                    </group>
```

rousseldenis replied:
@hailangvn This is reformated through pre-commit.

hailangvn replied:
Thanks @rousseldenis. I was new at the time I commented. :)

hailangvn commented on sale_promotion_rule/demo/product.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+    <record id="coupon" model="product.product">
+        <field name="name">Coupon</field>
+        <field name="type">service</field>
+    </record>
+</odoo>
```
Is this needed with current coupon logic?

hailangvn reviewed (approved):
Thanks for the PR. There are minor improvements not required to merge this.

hparfr commented:
Why not. I was not aware of this coupon module.

hailangvn commented:
My comment was about demo coupon product, which IMHO helps new user to understand this module. I was wondering what to do next after having a coupon product while we already added coupon code to promotions page. There is another demo file as well.

rousseldenis commented on sale_promotion_rule/views/sale_promotion_rule.xml:
```diff
@@ -0,0 +1,96 @@
+                            <field name="multi_rule_strategy" />
+                        </group>
+                        <group name="condition" string="Condition" col="3">
+                            <label for="date_from" string="Date début/fin" />
```
Could you change this to English version ?

rousseldenis replied:
@xavier-bouquiaux

sebastienbeau replied:
ping @xavier-bouquiaux

francesco-ooops commented:
@xavier-bouquiaux @rousseldenis wouldn't it make more sense to move this module to sale-promotion repo?

let us know if you need us to work on it

francesco-ooops commented:
or rather: what are the differences between this module's promotions and odoo standard promotions? should they coexist?

hparfr commented:
>or rather: what are the differences between this module's promotions and odoo standard promotions? should they coexist?

After few months of implementation with coupons in v14, it's missing key features for our e-commerce case (like apply more than 1 promotion at time).  Coupon is not designed to allow behavior changes. So it's a bit hacky to add features on top of it. We spent a lot of effort to make it work. 

Sale promotion rule, is an old module date here from 2017 and I think it came from a time when we had to synchronize with Magento / Prestashop and cover main features. It's designed and implemented to allow behavior changes if needed. I think a good refactoring will not hurt.

Like always, it depends of your needs and your budget.

rousseldenis commented:
@Cedric-Pigeon

rousseldenis commented:
> After few months of implementation with coupons in v14, it's missing key features for our e-commerce case (like apply more than 1 promotion at time). Coupon is not designed to allow behavior changes. So it's a bit hacky to add features on top of it. We spent a lot of effort to make it work.

@hparfr  But from 13 version, it is possible to use coupon for its model only and build on top of it as we did here : https://github.com/acsone/acsone-addons/tree/14.0/account_wallet_coupon

sebastienbeau reviewed (approved):
LGTM (code review), we should merge it even odoo have an alternative implementation (coupon do not solve all case). In a long terme we should see if we can use coupon or improve this module. 
For now some customer need such feature so let's merge it

xavier-bouquiaux committed:
[IMP] sale_promotion_rule : translate field

xavier-bouquiaux commented:
> LGTM (code review), we should merge it even odoo have an alternative implementation (coupon do not solve all case). In a long terme we should see if we can use coupon or improve this module. For now some customer need such feature so let's merge it

Changed to english version and rebased

regards

xbo

xavier-bouquiaux committed:
[IMP] pre-commit update

sebastienbeau commented:
/ocabot merge patch
